File: MiddleClick/TrayMenu.m
#import "TrayMenu.h"
#import "PreferenceKeys.h"
#import "Controller.h"
#import <Cocoa/Cocoa.h>

@implementation TrayMenu

- (id)initWithController:(Controller*)ctrl
{
  [super init];
  myController = ctrl;
  [self setChecks];
  return self;
}

- (void)initAccessibilityPermissionStatus:(NSMenu*)menu
{
  BOOL hasAccessibilityPermission = AXIsProcessTrusted();

  [self updateAccessibilityPermissionStatus:menu
                 hasAccessibilityPermission:hasAccessibilityPermission];

  if (!hasAccessibilityPermission) {
    [NSTimer scheduledTimerWithTimeInterval:0.3 repeats:NO block:^(NSTimer* timer) {
      [self initAccessibilityPermissionStatus:menu];
    }];
  }
}
- (void)updateAccessibilityPermissionStatus:(NSMenu*)menu
                 hasAccessibilityPermission:(BOOL)isTrusted
{
  _statusItem.button.appearsDisabled = !isTrusted;
  accessibilityPermissionStatusItem.hidden = isTrusted;
  accessibilityPermissionActionItem.hidden = isTrusted;
}

- (void)openWebsite:(id)sender
{
  NSURL* url = [NSURL
                URLWithString:@"https://github.com/artginzburg/MiddleClick-Sonoma"];
  [[NSWorkspace sharedWorkspace] openURL:url];
}
- (void)openAccessibilitySettings:(id)sender
{
  BOOL isPreCatalina = (floor(NSAppKitVersionNumber) < NSAppKitVersionNumber10_15);
  if (isPreCatalina) {
    NSAppleScript *a = [[NSAppleScript alloc] initWithSource:
                        @"tell application \"System Preferences\"\n"
                        "activate\n"
                        "reveal anchor \"Privacy_Accessibility\" of pane \"com.apple.preference.security\"\n"
                        "end tell"];
    [a executeAndReturnError:nil];
    [a release];
  } else {
    NSURL* url = [NSURL
                  URLWithString:@"x-apple.systempreferences:com.apple.preference.security?Privacy_Accessibility"];
    [[NSWorkspace sharedWorkspace] openURL:url];
  }
}

- (void)toggleTapToClick:(id)sender
{
  [myController setMode:[sender state] == NSControlStateValueOn];
  [self setChecks];
}

- (void)resetTapToClick:(id)sender
{
  [myController resetClickMode];
  [self setChecks];
}

- (void)setChecks
{
  bool clickMode = [myController getClickMode];
  NSString* clickModeInfo = clickMode ? @"Click" : @"Click or Tap";
  
  int fingersQua = (int)[[NSUserDefaults standardUserDefaults] integerForKey:kFingersNum];
  
  [infoItem setTitle:[clickModeInfo stringByAppendingFormat: @" with %d Fingers", fingersQua]];
  [tapToClickItem setState:clickMode ? NSControlStateValueOff : NSControlStateValueOn];
}

- (void)actionQuit:(id)sender
{
  [NSApp terminate:sender];
}

- (NSMenu*)createMenu
{
  NSMenu* menu = [NSMenu new];
  NSMenuItem* menuItem;
  
  
  
  [self createMenuAccessibilityPermissionItems:menu];
  
  // Add About
  menuItem = [menu addItemWithTitle:[NSString stringWithFormat:@"About %@...", getAppName()]
                             action:@selector(openWebsite:)
                      keyEquivalent:@""];
  [menuItem setTarget:self];
  
  [menu addItem:[NSMenuItem separatorItem]];
  
  infoItem = [menu addItemWithTitle:@""
                             action:nil
                      keyEquivalent:@""];
  [infoItem setTarget:self];
  
  tapToClickItem = [menu addItemWithTitle:@"Tap to click"
                                   action:@selector(toggleTapToClick:)
                            keyEquivalent:@""];
  [tapToClickItem setTarget:self];
  
  NSMenuItem* resetItem = [menu addItemWithTitle:@"Reset to System Settings"
                                          action:@selector(resetTapToClick:)
                                   keyEquivalent:@""];
  resetItem.alternate = YES;
  resetItem.keyEquivalentModifierMask = NSEventModifierFlagOption;
  [resetItem setTarget:self];
  
  [self setChecks];
  
  // Add Separator
  [menu addItem:[NSMenuItem separatorItem]];
  
  // Add Quit Action
  menuItem = [menu addItemWithTitle:@"Quit"
                             action:@selector(actionQuit:)
                      keyEquivalent:@"q"];
  [menuItem setTarget:self];
  
  return menu;
}

- (void)createMenuAccessibilityPermissionItems:(NSMenu *)menu
{
  accessibilityPermissionStatusItem = [menu addItemWithTitle:@"Missing Accessibility permission" action:nil keyEquivalent:@""];
  accessibilityPermissionActionItem = [menu addItemWithTitle:@"Open Privacy Preferences" action:@selector(openAccessibilitySettings:) keyEquivalent:@","];

  [menu addItem:[NSMenuItem separatorItem]];
}

- (void)applicationDidFinishLaunching:(NSNotification*)notification
{
  NSMenu* menu = [self createMenu];
  
  NSImage* icon = [NSImage imageNamed:(@"StatusIcon")];
  [icon setSize:CGSizeMake(24, 24)];
  
  // Check if Darkmode menubar is supported and enable templating of the icon in
  // that case.
  
  BOOL oldBusted = (floor(NSAppKitVersionNumber) <= NSAppKitVersionNumber10_9);
  if (!oldBusted) {
    // 10.10 or higher, so setTemplate: is safe
    [icon setTemplate:YES];
  }
  
  _statusItem = [[[NSStatusBar systemStatusBar]
                  statusItemWithLength:24] retain];
  _statusItem.behavior = NSStatusItemBehaviorRemovalAllowed;
  _statusItem.menu = menu;
  _statusItem.button.toolTip = getAppName();
  _statusItem.button.image = icon;

  [self initAccessibilityPermissionStatus:menu];
  
  [menu release];
}

- (BOOL)applicationShouldHandleReopen:(NSApplication *)sender
                    hasVisibleWindows:(BOOL)flag
{
  _statusItem.visible = true;
  return 1;
}

NSString* getAppName(void) {
    return [[NSProcessInfo processInfo] processName];
}

@end

File: MiddleClick/TrayMenu.swift
import Cocoa

@MainActor class TrayMenu: NSObject, NSApplicationDelegate {
  let myController: Controller
  var infoItem: NSMenuItem!
  var tapToClickItem: NSMenuItem!
  var accessibilityPermissionStatusItem: NSMenuItem!
  var accessibilityPermissionActionItem: NSMenuItem!
  var statusItem: NSStatusItem!

  init(controller: Controller) {
    myController = controller
    super.init()
  }

  // Initialize accessibility permission status
  @objc func initAccessibilityPermissionStatus(menu: NSMenu) {
    let hasAccessibilityPermission = AXIsProcessTrusted()

    updateAccessibilityPermissionStatus(
      menu: menu, hasAccessibilityPermission: hasAccessibilityPermission)

    if !hasAccessibilityPermission {
      Timer
        .scheduledTimer(
          timeInterval: 0.3,
          target: self,
          selector: #selector(initAccessibilityPermissionStatus(menu:)),
          userInfo: nil,
          repeats: false
        )
    }
  }

  // Update accessibility permission status
  func updateAccessibilityPermissionStatus(menu: NSMenu, hasAccessibilityPermission: Bool) {
    statusItem.button?.appearsDisabled = !hasAccessibilityPermission
    accessibilityPermissionStatusItem.isHidden = hasAccessibilityPermission
    accessibilityPermissionActionItem.isHidden = hasAccessibilityPermission
  }

  // Open the website
  @objc func openWebsite(sender: Any) {
    if let url = URL(string: "https://github.com/artginzburg/MiddleClick-Sonoma") {
      NSWorkspace.shared.open(url)
    }
  }

  // Open accessibility settings
  @objc func openAccessibilitySettings(sender: Any) {
    let isPreCatalina =
      (floor(NSAppKitVersion.current.rawValue) < NSAppKitVersion.macOS10_15.rawValue)
    if isPreCatalina {
      let appleScript = """
        tell application "System Preferences"
        activate
        reveal anchor "Privacy_Accessibility" of pane "com.apple.preference.security"
        end tell
        """
      if let script = NSAppleScript(source: appleScript) {
        script.executeAndReturnError(nil)
      }
    } else {
      if let url = URL(
        string: "x-apple.systempreferences:com.apple.preference.security?Privacy_Accessibility")
      {
        NSWorkspace.shared.open(url)
      }
    }
  }

  // Toggle Tap to Click
  @objc func toggleTapToClick(sender: NSButton) {
    myController.setMode(sender.state == .on)
    setChecks()
  }

  // Reset Tap to Click
  @objc func resetTapToClick(sender: NSButton) {
    myController.resetClickMode()
    setChecks()
  }

  // Set the checks based on configuration
  func setChecks() {
    let clickMode = myController.getClickMode()
    let clickModeInfo = "Click" + (clickMode ? "" : " or Tap")

    let fingersQua = UserDefaults.standard.integer(forKey: MiddleClickConfig.fingersNumKey)
    let allowMoreFingers = UserDefaults.standard.bool(forKey: MiddleClickConfig.allowMoreFingersKey)
    let fingersInfo = " with \(fingersQua)\(allowMoreFingers ? "+" : "") Fingers"

    infoItem.title = clickModeInfo + fingersInfo
    tapToClickItem.state = clickMode ? .off : .on
  }

  // Quit action
  @objc func actionQuit(sender: Any) {
    NSApp.terminate(sender)
  }

  // Create the menu
  func createMenu() -> NSMenu {
    let menu = NSMenu()

    createMenuAccessibilityPermissionItems(menu: menu)

    // Add About
    let aboutItem = menu.addItem(
      withTitle: "About \(getAppName())...", action: #selector(openWebsite(sender:)),
      keyEquivalent: "")
    aboutItem.target = self

    menu.addItem(NSMenuItem.separator())

    // Add info item
    infoItem = menu.addItem(withTitle: "", action: nil, keyEquivalent: "")
    infoItem.target = self

    // Add Tap to Click
    tapToClickItem = menu.addItem(
      withTitle: "Tap to click", action: #selector(toggleTapToClick), keyEquivalent: "")
    tapToClickItem.target = self

    // Add Reset
    let resetItem = menu.addItem(
      withTitle: "Reset to System Settings", action: #selector(resetTapToClick(sender:)),
      keyEquivalent: "")
    resetItem.isAlternate = true
    resetItem.keyEquivalentModifierMask = .option
    resetItem.target = self

    setChecks()

    // Add Separator
    menu.addItem(NSMenuItem.separator())

    // Add Quit
    let quitItem = menu.addItem(
      withTitle: "Quit", action: #selector(actionQuit(sender:)), keyEquivalent: "q")
    quitItem.target = self

    return menu
  }

  func createMenuAccessibilityPermissionItems(menu: NSMenu) {
    accessibilityPermissionStatusItem = menu.addItem(
      withTitle: "Missing Accessibility permission", action: nil, keyEquivalent: "")
    accessibilityPermissionActionItem = menu.addItem(
      withTitle: "Open Privacy Preferences", action: #selector(openAccessibilitySettings(sender:)),
      keyEquivalent: ",")
    menu.addItem(NSMenuItem.separator())
  }

  func getAppName() -> String {
    return ProcessInfo.processInfo.processName
  }

  func applicationDidFinishLaunching(_ notification: Notification) {
    let menu = createMenu()

    let icon = NSImage(named: "StatusIcon") ?? NSImage()
    icon.size = CGSize(width: 24, height: 24)  // TODO? increase size

    let oldBusted = (floor(NSAppKitVersion.current.rawValue) <= NSAppKitVersion.macOS10_9.rawValue)
    if !oldBusted {
      icon.isTemplate = true
    }

    statusItem = NSStatusBar.system.statusItem(withLength: NSStatusItem.squareLength)
    statusItem.behavior = .removalAllowed
    statusItem.menu = menu
    statusItem.button?.toolTip = getAppName()
    statusItem.button?.image = icon

    initAccessibilityPermissionStatus(menu: menu)
  }

  func applicationShouldHandleReopen(_ sender: NSApplication, hasVisibleWindows flag: Bool) -> Bool
  {
    statusItem.isVisible = true
    return true
  }
}
